Support For Stellaris Launchpad LM4F120

rakendrathapa opened the pull request:
This code checkin contains the support for Stellaris Launchpad LM4F120 board. 
I have worked specifically with ek-lm4f120xl board and it passes all the basic sanity tests.

- Stellaris Launchpad board EK-LM4F120XL contains Stellaris 64-pin LM4F120H5QR digital signal controller, which is a 80MHz 32-bit ARM Cortex M4 based MCU with floating point
- 256KM Flash/32KM SRAM / 2KB EEPROM on chip memory.

From personal experience of working on it, Its a great board which provides added advantage of compactness, extensible and flexibility owing to the feature of plug-in modules called BoosterPacks that fit on top of LaunchPad.

![ek-lm4f120xl](https://cloud.githubusercontent.com/assets/4710632/7673203/09bb78a8-fd2c-11e4-8b42-7884d0c11f45.jpg)


The major directories that have been checked in are:
boards/ek-lm4f120xl - board specific file
cpu/StellarisWare - Support Library for Stellaris Launchpad boards.
cpu/lm4f120 - cpu specific files

- cpu/StellarisWare library that is available for download by TI and supports all the Stellaris Launchpad board. In this checkin I have stripped down this library to the set that is useful for us. 
I have kept this under cpu and can be used by other TI Stellaris controllers for easy Porting of RIOT-OS. Though subsequently, I plan to replace this library with common routines to be used by any Stellaris Board.  

- Other directories are as per the norm. 

Currently, the board supports only the minimal set of functionality. Currently supported peripherals are : UART, TIMERS and ADC. 
Still other peripherals are yet to be implemented.

** Major ToDo List that remains:
- Adding other peripherals. 
- Extending existing peripherals. 
- Common library routines. 

Please let me know if other details are required. 

Regards,
Rakendra Thapa

kaspar030 commented on boards/ek-lm4f120xl/board.c:
```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright (C) 2014 Freie Universität Berlin
```
copyright?

rakendrathapa replied:
Oh..!! I copied Copyright note from the other existing files. 
I will edit this in accordance to the Coding Convention.

Thanks 
Rakendra

kaspar030 commented:
Could you split the commits for StellarisWare and the RIOT integration? Like this it's impossible to review.

haukepetersen commented:
Hi, welcome to RIOT and thanks for your PR! As far as I can tell, your code compiles fine. But I see one pressing issues that need to be fixed before we can review this PR in depth: 

We can not include the StellarisWare into RIOT. Just have a look at this snip from their license header:
```
// Texas Instruments (TI) is supplying this software for use solely and
// exclusively on TI's microcontroller products. The software is owned by
// TI and/or its suppliers, and is protected under applicable copyright
// laws. You may not combine this software with "viral" open-source
// software in order to form a larger program.
```
If I am not completely mistaken, RIOT is "viral" open-source software, right?

Further I would argue against including this hole firmware for more reasons:

- it is huge in terms of lines of code (reviewing your the 'actual' cpu/board code in github is impossible...)
- from experience it is more efficient to code directly against the CMSIS/register define headers
- the code contains quite some things that are not needed in RIOT

So I see two choices: (i) we drop the StellarisWare and implement the CPU code directly on CMSIS headers (even if we have to write parts of them ourselves) or (ii) include the StallarisWare as external packet in `pkt`. I would strongly opt for (i).


Please also have a look at the RIOT coding conventions [here](https://github.com/RIOT-OS/RIOT/wiki/Coding-conventions). Especially the usage of spaces instead of tabs and the naming scheme for variables/functions...

So lets get this thing right and I agree, the board is indeed an interesting platform!

rakendrathapa commented:
Thank you Kaspar and Hauke for prompt review feedback. 

- Ok. I had checked the license of StellarisWare and it said that it is license free and Royalty free and thus I thought I could include this in main code stream. The library used a BSD license and some example files which were not BSD licensed, I removed them from the checkin. 
But, as you have highlighted that it cant be included with "viral" open-source software, I will work towards removing/moving it.

- Regarding the second point of whether of porting StellarisWare or not as you have mentioned., I was in this opinion. 
Actually, one important advantage of using StellarisWare library with Stellaris MCU is that it's also included in the ROM inside the LM4F120 MCU  which helps saving Flash memory space.
Now, for the starters if we include this library as external packet, this will help getting this board quickly up and working and we can start testing applications on RIOT. 
Subsequently, I do plan to completely replace dependency on this library with our own defined header files and APIs. But, currently my approach was like:
- Complete Porting of RIOT-OS. 
- Gradually replace StellarisWare. 

So, let me know your view on this. I have already started working on it, so it should not be a major problem either way. 

In case we go with option ii)  , I will commit StellarisWare in a separate ER for review. 


Thanks and Regards,
Rakendra

kaspar030 commented:
IMO adding the library as external package would be OK, so going this route for now and then working towards removing the dependency would be fine with me. @haukepetersen?

haukepetersen commented:
Yap, fine with it. The fact that the library code is already included on a ROM on in the CPU is new to me, so under that condition it makes definitely sense to use this code - so I would expect that we can strip down the StellaWare to the header files defining these included functions?!

rakendrathapa commented:
Yes, exactly. We can strip down StellarisWare library to just the header files. I re-checked the code, there are some unnecessary project files and .o .d etc files in the checked in StellarisWare library code. I will clean it up as well and add it as external package.

rakendrathapa commented:
Hi,

Please check the new commits that I have made. 

- There is a separate commit for StellarisWare library header file and the board support files.
- I have modified the cpu/board files to refer the library code in the ROM. It actually gives an significant decrease in size of upto 6K. :)

Also, I have stripped Stellarisware library to just the Header files. I have removed the header files that had usage restriction since anyways we were not using this. So, I think its okay to merge it with main code. 

Thanks and Regards,
Rakendra

haukepetersen commented:
Hi, you commits do still contain an awefull lot of code (90k lines while RIOT all together has ~250K...). Here is what I would suggest:

From the 2.3 million lines (counted using `wc -l`) in `StellarisWare/driverlib` and `StellarisWare/inc` are only 100K lines of headers for peripherals and the ROM headers. The rest are CPU specific headers for every single TI MCU in that family. From second group we surely do not need to include all of them. So only leave 1! of them (the one for the MCU on the board you are porting) and remove the rest. We can add them whenever there is another stellaris MCU we add to RIOT. And this is exactly how we handle it with all the other MCU families in RIOT.

Regarding the folder structure: Please make it consistent to all the other platforms in RIOT. I would do it as follows:
```
cpu
 | - stellaris_common
 |    | - include
 |         | - asmdefs.h
 |         | - hw_*.h
 |         | - stellaris_lib
 |              | - adc.h
 |              | - can.h
 |              | - ...
 | - lm4f210
      | - include
      |    | - lm4f120h5qr.h
      |    | - cpu-conf.h etc.
      | - lm4f120h5qr_linkerscript.ld
      | - ...
```

One further structural thing: please make use of the cortex-m4_common code!

rakendrathapa commented:
Okay. I will work on this further library stripping and structuring and will make commits soon. 
Thanks for feedback.

rakendrathapa commented:
Hi,

Please check the latest commits for lm4f120.
I have modified the repository with the suggested major changes:

- Stripped the StellarisWare library further with just the header files required by the lm4fh5qr cpu. 

- Restructuring -
Restructured the stellarisware library as suggested to make it consistent to other platforms in RIOT. 

cpu
 | - stellaris_common
 |    | - include
 |         | - asmdefs.h
 |         | - hw_*.h
 |         | - stellaris_periph
 |              | - adc.h
 |              | - can.h
 |              | - ...
 |
 | - lm4f210
      | - include
      |    | - lm4f120h5qr.h
      |    | - cpu-conf.h etc.
      | - lm4f120h5qr_linkerscript.ld
      | - ...


- Included dependency on cortex_m4_common model. 
- Done major sanity tests. Should be working fine. 

Please lemme know the review feedback and if anything further needs to be amended.

Thanks and Regards,
Rakendra

gebart commented:
periph/adc needs some love.

	Building application: periph_adc (43/81) .....
		failed .. retrying: ek-lm4f120xl
		success: msbiot, mulle, pba-d-01-kw2x, stm32f4discovery

rakendrathapa commented:
:+1:  Given..

PeterKietzmann commented:
Anyone willing to be the new assignee for this PR? I think I won't get around to go into details here in the next time. Also I don't even know the hardware or have a test-device within reach.

haukepetersen commented:
I just got my hands on one of those boards, so I can test this.

haukepetersen commented:
Just did a quick test, looks good so far. Here are some issues I have found:

 - I think you forgot to add the openocd config file. I just [PRd it against your branch](https://github.com/rakendrathapa/RIOT/pull/1)
 - `hello-world` works fine
 - `default` and `test/periph_uart_int` do not, seems like there is trouble with the implementation of the UART interrupts

The `stellaris_common` part now looks much slimmer and cleaner! There is still one thing that could be optimized: The file `cpu/stellaris_common/include/stellaris_periph/pin_map.h` is still VERY large, as it contains the pin mappings for all available stellaris MCUs. How about removing it and copying just the part needed for the `LM4F120H5QR` CPU into a new file `cpu/lm4f120/include/lm4f120h5qr_pin_map.h`? Additional pin mappings can be added as needed, and we don't have to maintain a 60K+ line header file...

rakendrathapa commented:
@haukepetersen .. Okay ..I will modify pin_map file. I was working on this for sometime and have made the file already so this checkin will be done soon. 
- Okay even I just saw the error with test/periph_uart_int .. I think, interrupts are not being handled properly. May have to check other peripheral interrupts as well.  
Thanks for reviewing this.

rakendrathapa commented:
By the way, I needed some help here as well.
- How are you running the test cases for the board? 
I mean using test script to run each test case one by one OR testing each case individually to see the output.

haukepetersen commented:
Hej, at the moment we run the test-cases manually (and look at the output). Though we are planning on some more automated infrastructure for the future, we have not made the progress we wanted to in that direction...

gebart commented:
@rakendrathapa you need to fix the whitespace errors, run `git diff --check master` or `dist/tools/whitespacecheck/check.sh` to see the errors.

rakendrathapa commented:
Ok. Will do that.
Is there also buildin script to replace tab by spaces. ?

gebart commented:
@rakendrathapa I usually use either the `expand` Unix command or the "Replace tabs by spaces" menu option in my editor (Geany)

OlegHahm commented:
The corresponding settings in vim are:
```
set tabstop=4
set expandtab
```
and the `retab` command.

haukepetersen commented:
your PR needs a rebase. Please have also a look at #2852 and adapt your code accordingly (it should actually make your code/makefiles simpler...).

rakendrathapa commented:
Thanks for the update @haukepetersen.
I am out on vacation for this week and have very limited internet access. I will update the repository with the suggested changes as soon as I am back.

haukepetersen commented:
no problem, just update the branch whenever you are back. By then we should have also a lot of simplifications in the common code in place to make your life easier :-)

rakendrathapa commented:
Hi, 

I have modified the files as desired. But then there are lot of merging conflicts have have cropped in with the recent changes. Since, I had done the changes on top of the parent stream, I will be deleting this repository and creating a new one. 

Thanks and Regards,
Rakendra

kaspar030 commented:
Thanks a lot for going through the trouble of creating a nice port!
